module_utils/certificate/providers/base.py

```diff
@@ -8,12 +8,78 @@
 from cryptography.hazmat.backends import default_backend
 from cryptography.x509.oid import NameOID
 
+from pyasn1.codec.der import decoder
+from pyasn1.type import char, namedtype, tag, univ
+
 from ansible.module_utils import six
 
 if six.PY2:
     FileNotFoundError = IOError  # pylint: disable=redefined-builtin
 
 
+class _PrincipalName(univ.Sequence):
+    componentType = namedtype.NamedTypes(
+        namedtype.NamedType(
+            "name-type",
+            univ.Integer().subtype(
+                explicitTag=tag.Tag(tag.tagClassContext, tag.tagFormatSimple, 0)
+            ),
+        ),
+        namedtype.NamedType(
+            "name-string",
+            univ.SequenceOf(char.GeneralString()).subtype(
+                explicitTag=tag.Tag(tag.tagClassContext, tag.tagFormatSimple, 1)
+            ),
+        ),
+    )
+
+
+class _KRB5PrincipalName(univ.Sequence):
+    componentType = namedtype.NamedTypes(
+        namedtype.NamedType(
+            "realm",
+            char.GeneralString().subtype(
+                explicitTag=tag.Tag(tag.tagClassContext, tag.tagFormatSimple, 0)
+            ),
+        ),
+        namedtype.NamedType(
+            "principalName",
+            _PrincipalName().subtype(
+                explicitTag=tag.Tag(tag.tagClassContext, tag.tagFormatSimple, 1)
+            ),
+        ),
+    )
+
+
+class KRB5PrincipalName(x509.OtherName):
+    """Kerberos Principal x509 OtherName implementation."""
+
+    # pylint: disable=too-few-public-methods
+
+    oid = "1.3.6.1.5.2.2"
+
+    def __init__(self, type_id, value):
+        super(KRB5PrincipalName, self).__init__(type_id, value)
+        self.name = self._decode_krb5principalname(value)
+
+    @staticmethod
+    def _decode_krb5principalname(data):
+        principal = decoder.decode(data, asn1Spec=_KRB5PrincipalName())[0]
+        realm = six.ensure_text(
+            str(principal["realm"]).replace("\\", "\\\\").replace("@", "\\@")
+        )
+        name = principal["principalName"]["name-string"]
+        name = u"/".join(
+            six.ensure_text(str(n))
+            .replace("\\", "\\\\")
+            .replace("/", "\\/")
+            .replace("@", "\\@")
+            for n in name
+        )
+        name = u"%s@%s" % (name, realm)
+        return name
+
+
```

Comment:
@rcritten - does this look ok to you?

Reply:
It might be good if @tiran can also take a look.

Reply:
IPA has a class for handling principals. I'm not sure if this code is based on that or home grown. It would be nice if the IPA code could be used.

Reply:
@rcritten the implementation it's heavily based on IPA x509 code.

Reply:
We could just install and use ipalib.x509 instead of implementing this here but that will add 126 package dependencies to linux-system-roles [1]. Since the code is not that large I decided to go for the local implementation. @richm @pcahyna what do you think?

[1] https://gist.github.com/seocam/a1bb35dbc68feff90ea08a565d35f9c1

Reply:
If using the `certmonger` provider, will the system already have the `ipalib.x509` package installed?  If not, will installing it pull in a lot of unnecessary stuff?

Reply:
No, `certmonger` does not depend on `ipalib`.

Reply:
I'm ok with this copy/paste implementation, as it is a standard and unlikely to change, and doesn't seem to be particularly security sensitive.  If someone has a strong objection, please comment.

Reply:
19 extra packages on CentOS 8.

Reply:
> 19 extra packages on CentOS 8.

Doesn't seem worth it

Reply:
+1, it's ok with me.